test/onnx/test_pytorch_onnx_shape_inference.py

```diff
@@ -268,6 +271,60 @@ def test_resize_after_concat(self):
         )
         self.run_test(g, resize.node(), expect_tensor("Float", shape=(4, 32, 128, 128)))
 
+    def test_autograd_layernorm_shape(self):
+        def symbolic_pythonop(ctx: torch.onnx.SymbolicContext, g, *args, **kwargs):
+            return g.op("com.microsoft::PythonOp")
+
+        torch.onnx.register_custom_op_symbolic("prim::PythonOp", symbolic_pythonop, 1)
+        self.addCleanup(torch.onnx.unregister_custom_op_symbolic, "prim::PythonOp", 1)
+
+        # necessay parameters for transformer embeddings
+        hidden_size, max_position_embeddings, batch_size = (
+            48,
+            32,
+            2,
+        )
+
+        # wrap nn.layernorm into autograd.func repro the same
+        # issue as using apex fusedlayernorm
+        class CustomLayerNorm(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, embedding):
+                layer_norm = torch.nn.LayerNorm(hidden_size, eps=1e-12)
+                return layer_norm(embedding)
+
+        class EmbeddingModule(torch.nn.Module):
+            def forward(
+                self,
+                embeddings=None,
+            ):
+                embedding_output = CustomLayerNorm.apply(embeddings)
+                query = embedding_output.transpose(0, 1)
+                tgt_len, bsz, embed_dim = query.size()
+                query = query.reshape(tgt_len, bsz, embed_dim)
+                return query
+
+        embeddings = torch.randn(batch_size, max_position_embeddings, hidden_size)
+
+        f = io.BytesIO()
+        torch.onnx.export(
+            EmbeddingModule().eval(),
+            (embeddings),
+            f,
+            opset_version=self.opset_version,
+            input_names=["embeddings"],
+            dynamic_axes={"embeddings": [0, 1, 2]},
+            custom_opsets={"com.microsoft": 1},
+        )
+        graph = onnx.load(io.BytesIO(f.getvalue()))
+
+        # If there is a onnx::Constant node with dim=3, it is making the shape static
+        dim = 3
+        for node in graph.graph.node:
```

Comment:
How can we make sure the for loop is run at least once and the test enters the if branches?

Reply:
This is a good point. I think the test is built with precondition that the graph is at least functional. It would be super broken for a graph to not enter the if branch, and if that's the case, this test wouldn't be the most guilt one?

Reply:
one possibility: Maybe a simple boolean var that is set inside the branch will do.

Reply:
Within the new commit, I am taking out Constant node first, and check if their values being static or not. I think now is more clear and less ambiguous.

Reply:
Thanks!